google/cloud/storage/internal/async/reader_connection_resume.cc

```diff
@@ -0,0 +1,96 @@
+// Copyright 2024 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "google/cloud/storage/internal/async/reader_connection_resume.h"
+
+namespace google {
+namespace cloud {
+namespace storage_internal {
+GOOGLE_CLOUD_CPP_INLINE_NAMESPACE_BEGIN
+
+using ::google::cloud::storage_experimental::AsyncReaderConnection;
+using ::google::cloud::storage_experimental::ResumePolicy;
+using ReadResponse =
+    google::cloud::storage_experimental::AsyncReaderConnection::ReadResponse;
+
+void AsyncReaderConnectionResume::Cancel() {
+  if (auto impl = CurrentImpl()) return impl->Cancel();
+}
+
+future<ReadResponse> AsyncReaderConnectionResume::Read() {
+  return Read(std::unique_lock<std::mutex>(mu_));
+}
+
+RpcMetadata AsyncReaderConnectionResume::GetRequestMetadata() {
+  if (auto impl = CurrentImpl()) return impl->GetRequestMetadata();
+  return {};
+}
+
+future<ReadResponse> AsyncReaderConnectionResume::Read(
+    std::unique_lock<std::mutex> lk) {
+  auto impl = CurrentImpl(lk);
+  lk.unlock();
+  if (!impl) return Reconnect();
+  return impl->Read().then([this](auto f) { return OnRead(f.get()); });
+}
+
+future<ReadResponse> AsyncReaderConnectionResume::OnRead(ReadResponse r) {
+  if (absl::holds_alternative<storage_experimental::ReadPayload>(r)) {
+    auto const& response = absl::get<storage_experimental::ReadPayload>(r);
+    received_bytes_ += response.size();
+    if (response.metadata().has_value() && !generation_.has_value()) {
+      generation_ = storage::Generation(response.metadata()->generation());
+    }
+    return make_ready_future(std::move(r));
+  }
+  auto const& status = absl::get<Status>(r);
+  if (status.ok() || resume_policy_->OnFinish(status) == ResumePolicy::kStop) {
+    return make_ready_future(std::move(r));
+  }
+  return Reconnect();
+}
+
+future<ReadResponse> AsyncReaderConnectionResume::Reconnect() {
+  // TODO(coryan) - do we need a `self` here?
```

Comment:
This seems like something to determine now.

Reply:
Indeed, fixed the comment to explain why this is safe.